### PR TITLE
Bug 1217205 - Imagecompare: test_marketplace_download_execute_map_app.py fails

### PR DESCRIPTION
**Pull request by nojunpark**

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=2129deb3426c160cd3b50ec71c4b1042f454d6fb

File: tests/python/gaia-ui-tests/gaiatest/apps/homescreen/regions/confirm_install.py [hunk 1]
```diff
@@ -21,3 +21,4 @@ def tap_confirm(self):
                 *self._confirm_install_button_locator))
             Wait(self.marionette).until(expected.element_displayed(confirm))
             confirm.tap()
+            self.apps.switch_to_displayed_app()
```
Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/apps/homescreen/regions/confirm_install.py]: I think this should be part of the system app regions.

Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/apps/homescreen/regions/confirm_install.py]: The time.sleep(1) call on line 18 doesn't make sense to me, from reading that bug, it seems to be a failure in b2g desktop, which we don't run/support anymore. I guess we could just remove it.

File: tests/python/gaia-ui-tests/gaiatest/apps/marketplace/app.py [hunk 2]
```diff
@@ -70,6 +70,10 @@ def tap_install_button(self):
         self.root_element.find_element(*self._install_button_locator).tap()
         self.marionette.switch_to_frame()
 
+    def tap_open_app_button(self):
+        if self.root_element.find_element(*self._install_button_locator).text == 'Open app':
+            self.root_element.find_element(*self._install_button_locator).tap()
```
Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/apps/marketplace/app.py]: This doesn't seem like a good idea to me, the if call.
What happens when there is no 'Open app'?

File: tests/python/gaia-ui-tests/gaiatest/tests/graphics/test_marketplace_download_execute_map_app.py [hunk 3]
```diff
@@ -34,62 +34,48 @@ def test_install_and_execute_bing_map(self):
...
-
-        installed_app = homescreen.installed_app(app_name)
-        installed_app.tap_icon()
+        # Do a search again and get the element again.  Since first_result is a pageregion object,
+        # after the frame change done during the install the element will become stale and need to be recaptured.
+        results = marketplace.search(self.app_search)
+        first_result = results.search_results[0]
+        first_result.tap_open_app_button()
 
+        self.apps.switch_to_displayed_app()
```
Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/tests/graphics/test_marketplace_download_execute_map_app.py]: It seems better to me if this would be part of tap_open_app_button.

File: tests/python/gaia-ui-tests/gaiatest/tests/graphics/test_marketplace_download_execute_map_app.py [hunk 4]
```diff
@@ -34,62 +34,48 @@ def test_install_and_execute_bing_map(self):
...
         Wait(self.marionette).until(lambda m: m.title == self.app_title)
         bingmap = self.marionette.find_element(*self._map_locator)
         Wait(self.marionette).until(lambda m: bingmap.is_displayed())
```
Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/tests/graphics/test_marketplace_download_execute_map_app.py]: And perhaps these lines as well.

File: tests/python/gaia-ui-tests/gaiatest/tests/graphics/test_marketplace_download_execute_map_app.py [hunk 5]
```diff
@@ -34,62 +34,48 @@ def test_install_and_execute_bing_map(self):
 
         results = marketplace.search(self.app_search)
         first_result = results.search_results[0]
-        app_name = first_result.get_app_name()
         first_result.tap_install_button()
 
-        # Confirm the installation and wait for the app icon to be present
+        # Confirm the installation
```
Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/tests/graphics/test_marketplace_download_execute_map_app.py]: If you'd ditch the comment completely, I don't think anything of value would be lost.

File: tests/python/gaia-ui-tests/gaiatest/tests/graphics/test_marketplace_download_execute_map_app.py [hunk 6]
```diff
@@ -34,62 +34,48 @@ def test_install_and_execute_bing_map(self):
...
-
-        # Check that the icon of the app is on the homescreen
-        homescreen = Homescreen(self.marionette)
-        homescreen.wait_for_app_icon_present(app_name)
-
-        installed_app = homescreen.installed_app(app_name)
-        installed_app.tap_icon()
+        # Do a search again and get the element again.  Since first_result is a pageregion object,
+        # after the frame change done during the install the element will become stale and need to be recaptured.
+        results = marketplace.search(self.app_search)
```
Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/tests/graphics/test_marketplace_download_execute_map_app.py]: Perhaps you could do results = SearchResults(self.marionette) here and work from there?
It seems a shame that you have to reload the page just because of this bug.
I think you should also mention the issue that is going on here, namely https://bugzilla.mozilla.org/show_bug.cgi?id=1202246

File: tests/python/gaia-ui-tests/gaiatest/apps/homescreen/regions/confirm_install.py (hunk 1 above)
Comment (npark-mozilla) [inline tests/python/gaia-ui-tests/gaiatest/apps/homescreen/regions/confirm_install.py]: I think there should be a separate PR for moving the file in. I did remove the sleep call.

File: tests/python/gaia-ui-tests/gaiatest/tests/graphics/test_marketplace_download_execute_map_app.py (hunk 3 above)
Comment (npark-mozilla) [inline tests/python/gaia-ui-tests/gaiatest/tests/graphics/test_marketplace_download_execute_map_app.py]: true, changed

File: tests/python/gaia-ui-tests/gaiatest/tests/graphics/test_marketplace_download_execute_map_app.py (hunk 4 above)
Comment (npark-mozilla) [inline tests/python/gaia-ui-tests/gaiatest/tests/graphics/test_marketplace_download_execute_map_app.py]: replaced

File: tests/python/gaia-ui-tests/gaiatest/tests/graphics/test_marketplace_download_execute_map_app.py (hunk 6 above)
Comment (npark-mozilla) [inline tests/python/gaia-ui-tests/gaiatest/tests/graphics/test_marketplace_download_execute_map_app.py]: I found a way to got around it. thanks!

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=2787fa5a4c78f7ac48dc305b6d6c6420a356454e

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=69fc44edacc02bab5e71723b5830400c2ba9457e

File: tests/python/gaia-ui-tests/gaiatest/tests/graphics/test_marketplace_download_execute_map_app.py [hunk 7]
```diff
@@ -17,7 +16,7 @@ class TestSearchMarketplaceAndInstallApp(GaiaImageCompareTestCase):
     """
     app_search = 'Bing Maps :packaged'
     app_title = 'Bing Maps'
-    draw_wait_time = 5
+    draw_wait_time = 10
```
Comment (npark-mozilla) [inline tests/python/gaia-ui-tests/gaiatest/tests/graphics/test_marketplace_download_execute_map_app.py]: turns out 10 seconds is what it takes for the test to be reliable.

File: tests/python/gaia-ui-tests/gaiatest/apps/homescreen/regions/confirm_install.py (hunk 1 above)
Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/apps/homescreen/regions/confirm_install.py]: Sure, no problem.

File: tests/python/gaia-ui-tests/gaiatest/tests/graphics/test_marketplace_download_execute_map_app.py [hunk 8]
```diff
@@ -34,62 +33,39 @@ def test_install_and_execute_bing_map(self):
...
-
-        installed_app = homescreen.installed_app(app_name)
-        installed_app.tap_icon()
-
-        Wait(self.marionette).until(lambda m: m.title == self.app_title)
-        bingmap = self.marionette.find_element(*self._map_locator)
-        Wait(self.marionette).until(lambda m: bingmap.is_displayed())
-        time.sleep(self.draw_wait_time)
-        self.take_screenshot()
+        # Do a search again and get the element again.  Since first_result is a pageregion object,
```
Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/tests/graphics/test_marketplace_download_execute_map_app.py]: So the "Do a search again" part of the comment can be removed here then, right?

File: tests/python/gaia-ui-tests/gaiatest/apps/marketplace/app.py [hunk 9]
```diff
@@ -70,6 +78,15 @@ def tap_install_button(self):
         self.root_element.find_element(*self._install_button_locator).tap()
         self.marionette.switch_to_frame()
 
+    def tap_open_app_button(self, app_title, _app_locator):
+        button = self.root_element.find_element(*self._install_button_locator)
+        Wait(self.marionette).until(lambda m: button.text == 'Open app')
```
Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/apps/marketplace/app.py]: Perhaps this doesn't work in other locales, fwiw.

Comment (npark-mozilla) [inline tests/python/gaia-ui-tests/gaiatest/apps/marketplace/app.py]: yup, I don't think I saw [data-l10n-text] value for this element, so I had to check for the text, I'll double check though.  If there isn't one, then I think we should make this test @parameterized for each locale.

File: tests/python/gaia-ui-tests/gaiatest/tests/graphics/test_marketplace_download_execute_map_app.py (hunk 8 above)
Comment (npark-mozilla) [inline tests/python/gaia-ui-tests/gaiatest/tests/graphics/test_marketplace_download_execute_map_app.py]: right, will do

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=4040eeee01dd1f33c97a01b9f5e7cc861d971fa9
